File: core/nodesDev/gfToolsNodes/py/n_gfUtilAngleToDouble.py
# -*- coding: utf-8 -*-
"""
Copyright (c) 2019 Giuliano França

Redistribution:
    Something here.

Maya Node:
    [This is a prototype version of the gfUtilAngleToDouble node. You should be using the related C++ version.]
    This node is a test node and only performs test operations with one input value.

Requirements:
    Maya 2017 or above.

Todo:
    * Convert angle values to radians.

This code supports Pylint. Rc file in project.
"""
# pylint: disable=import-error
# import-error = Supress Maya modules import error

import maya.api.OpenMaya as om2


def maya_useNewAPI():
    """ Function to Maya recognize the use of the Python API 2.0. """
    # pylint: disable=invalid-name, unnecessary-pass
    pass


def INPUT_ATTR(FNATTR):
    """ Configure a input attribute. """
    # pylint: disable=invalid-name
    FNATTR.writable = True
    FNATTR.readable = True
    FNATTR.storable = True
    FNATTR.keyable = True


def OUTPUT_ATTR(FNATTR):
    """ Configure a output attribute. """
    # pylint: disable=invalid-name
    FNATTR.writable = False
    FNATTR.readable = True
    FNATTR.storable = False
    FNATTR.keyable = False


class AngleToDouble(om2.MPxNode):
    """ Main class of gfUtilAngleToDouble node. """

    kNODE_NAME = ""
    kNODE_CLASSIFY = ""
    kNODE_ID = ""

    inAngle = om2.MObject()
    outDouble = om2.MObject()

    def __init__(self):
        """ Constructor. """
        om2.MPxNode.__init__(self)

    @staticmethod
    def creator():
        """ Maya creator function. """
        return AngleToDouble()

    @staticmethod
    def initialize():
        """
        Defines the set of attributes for this node. The attributes declared in this function are assigned
        as static members to AngleToDouble class. Instances of AngleToDouble will use these attributes to create plugs
        for use in the compute() method.
        """
        uAttr = om2.MFnUnitAttribute()
        nAttr = om2.MFnNumericAttribute()

        AngleToDouble.inAngle = uAttr.create("angle", "angle", om2.MFnUnitAttribute.kAngle, 0.0)
        INPUT_ATTR(uAttr)

        AngleToDouble.outDouble = nAttr.create("outDouble", "od", om2.MFnNumericData.kDouble, 0.0)
        OUTPUT_ATTR(nAttr)

        AngleToDouble.addAttribute(AngleToDouble.inAngle)
        AngleToDouble.addAttribute(AngleToDouble.outDouble)
        AngleToDouble.attributeAffects(AngleToDouble.inAngle, AngleToDouble.outDouble)

    def compute(self, plug, dataBlock):
        """
        Node computation method:
            * plug is a connection point related to one of our node attributes (either an input or an output).
            * dataBlock contains the data on which we will base our computations.
        """
        # pylint: disable=no-self-use
        if plug != AngleToDouble.outDouble:
            return om2.kUnknownParameter

        angle = dataBlock.inputValue(AngleToDouble.inAngle).asAngle().asDegrees()
        outDoubleHandle = dataBlock.outputValue(AngleToDouble.outDouble)

        outDoubleHandle.setDouble(angle)
        outDoubleHandle.setClean()

File: core/plugin/__DEV/C++/source/n_gfUtilAngleToDouble.cpp
#include "n_gfUtilAngleToDouble.h"

// Configure a input attribute.
#define INPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(true);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(true);	\
	FNATTR.setKeyable(true);	\

// Configure a output attribute.
#define OUTPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(false);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(false);	\
	FNATTR.setKeyable(false);	\


// Constructor.
AngleToDouble::AngleToDouble() {}

// Destructor.
AngleToDouble::~AngleToDouble() {}

MObject AngleToDouble::inAngle;
MObject AngleToDouble::outDouble;

void* AngleToDouble::creator(){
    // Maya creator function.
    return new AngleToDouble();
}

MStatus AngleToDouble::initialize(){
    /*
    Defines the set of attributes for this node. The attributes declared in this function are assigned
    as static members to AngleToDouble class. Instances of AngleToDouble will use these attributes to create plugs
    for use in the compute() method.
    */
    MStatus status;
    MFnUnitAttribute uAttr;
    MFnNumericAttribute nAttr;

    inAngle = uAttr.create("angle", "angle", MFnUnitAttribute::kAngle, 0.0, &status);
    INPUT_ATTR(uAttr);

    outDouble = nAttr.create("outDouble", "od", MFnNumericData::kDouble, 0.0, &status);
    OUTPUT_ATTR(nAttr);

    addAttribute(inAngle);
    addAttribute(outDouble);
    attributeAffects(inAngle, outDouble);

    return status;
}

MStatus AngleToDouble::compute(const MPlug& plug, MDataBlock& dataBlock){
    /*
    Node computation method:
        * plug is a connection point related to one of our node attributes (either an input or an output).
        * dataBlock contains the data on which we will base our computations.
    */
    if (plug != outDouble)
        return MStatus::kUnknownParameter;

    double angle = dataBlock.inputValue(inAngle).asAngle().asDegrees();
    MDataHandle outDoubleHandle = dataBlock.outputValue(outDouble);

    outDoubleHandle.setDouble(angle);
    outDoubleHandle.setClean();

    return MStatus::kSuccess;
}